### ISSUE TITLE
How to: versioning, introspection

### ISSUE THREAD

**Issue by jensneuse** [reframejs/wildcard-api#15]
Very interesting project!

1. I'd like to know how a client might know which APIs are available, similar to GraphQL introspection.

2. How do you version/deprecate APIs? How would you ensure backwards compatibility.

Would be nice to have a FAQ that covers the full API lifecycle.

**Comment by brillout** [comment]
> I'd like to know how a client might know which APIs are available, similar to GraphQL introspection.

With Wildcard you create an internal API which means that you are the consumer of your own API. Since you (or your organization) wrote the API you already know what endpoints there are.

That said, the root route `/wildcard/` will list all your endpoints while developing (`[undefined, 'development'].includes(process.env.NODE_ENV)`) which is comfy for debugging.

> How do you version/deprecate APIs? How would you ensure backwards compatibility.

I would do this on an endpoint-by-endpoint basis. For example:

~~~js
endpoints.getDataForProductsPage_v1 = function({productId}) {
   /* ... */
};
endpoints.getDataForProductsPage_v2 = function({productId}) {
   /* ... */
};
~~~

Wildcard is primarily meant to be used with a web frontend + Node.js server. As long as you deploy both at the same time, you don't need any versioning.

There are three cases where you need versioning;
- On mobile where you have no control over client deployments.
- For an API consumed by third parties.
- For an API consumed internally but within a large organization.

I can't think of any use case beyond these three that need versioning.

Wildcard is meant to be used to create internal APIs and Wildcard doesn't support mobile. So my answer is: you most likely don't need versioning.

> Would be nice to have a FAQ that covers the full API lifecycle.

Good point.

> Very interesting project!

Danke :-).

Do you have ideas how I could popularize Wildcard?

**Comment by jensneuse** [comment]
Do you think this actually works across an organisation? If you have a 1:1 relationship between front-end and backend with the codebase in a mono repo I think this will work fine.
If on the other hand you want to access this API from multiple front ends sitting in different repositories I feel like this approach has its shortcomings in terms of documentation and discoverability.
Both REST (with OIDC) and GraphQL will be a better fit in these scenarios as they enable developers to easily discover APIs and learn about how to use them.
Maybe I'm missing something so please correct me if I'm wrong.

**Comment by brillout** [comment]
As far as I can see, the most important criteria here is how many API consumers you have.

For example, if you have only 2 front-ends consuming yourre- RPC API, then it's most likely acceptable to change the API every time one of these frontends needs a change.

But, if you have an API that has 1000s of different consumers, then RPC is not the right choice. You just cannot change the API every time one of these 1000s of consumers needs a change.

Note that [gRPC](https://grpc.io/) is used within large organizations such as Google, Netflix, Cisco or Square.


> this approach has its shortcomings in terms of documentation and discoverability

Yes. Auto doc generation is out of scope right now. Maybe I'll explore this in the future though.

**The future**

Things may change in the future. On a high level, RPC is essentially just a permission layer. Nothing hinders you to use Wildcard to create a generic API that is able to fulfill all kinds of data requirements. You could actually even create a GraphQL-like API with Wildcard. This is uncharted territory and it would be exciting to explore what could be done in that space. You could imagine ORMs that would automatically hook themselves onto Wildcard: you'd be able to use the ORM directly from the frontend. All you'd need to do is to define "permission functions" that basically define what users are allowed to read/write what objects.

But, right now, I'm focused on use cases where there is a 1:1 relationship between front-end and backend. Only after RPC & Wildcard gets popular will I expand the scope of this project.

**Comment by brillout** [comment]
Let me know if you have further questions or doubts.

**Comment by brillout** [comment]
I created a [FAQ](https://github.com/reframejs/wildcard-api/blob/master/docs/faq.md#faq) which answers the question "How can I do versioning with RPC?". It does a better job at explaining why you don't need versioning with RPC than what I wrote in this ticket.
